### PR TITLE
fips: disable TLS1.3 chacha20-poly1305

### PR DESCRIPTION
**Pull request by kyessenov**
Commit Message: Disables a cipher from hard-coded TLSv1.3 list in FIPS build 
Additional Description: Fixes https://github.com/envoyproxy/envoy/issues/31746 until CMVP certification of the next boringssl_fips version (ETA May). We may hold on to merging this, but this patch is likely useful for vendors.
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:

### CODE REVIEW

**Comment by kyessenov** [comment]
End-to-end test:
```
> openssl s_client -connect localhost:8081 -ciphersuites TLS_CHACHA20_POLY1305_SHA256 -CAfile testdata/certs/root.cert
CONNECTED(00000003)
40E77FBF207F0000:error:0A000410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:../ssl/record/rec_layer_s3.c:1586:SSL alert number 40
```

File: bazel/boringssl_fips.patch [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+index 7ad821041..1a816f48a 100644
+--- a/boringssl/ssl/s3_both.cc
++++ b/boringssl/ssl/s3_both.cc
+@@ -691,6 +691,16 @@ class CipherScorer {
+   const bool security_128_is_fine_;
+ };
+ 
++
++bool ssl_tls13_cipher_meets_policy(uint16_t cipher_id) {
++  switch (cipher_id) {
```
Comment (RyanTheOptimist) [inline bazel/boringssl_fips.patch:48]: nit: How about just: `return cipher_id != TLS1_CK_CHACHA20_POLY1305_SHA256 & 0xffff`

Comment (RyanTheOptimist) [reply bazel/boringssl_fips.patch:48]: (Or is this currently implementation helpful for future integration?)

File: bazel/external/boringssl_fips.genrule_cmd [hunk 2]
```diff
@@ -121,7 +121,8 @@ rm -rf boringssl/build
 cd boringssl
 mkdir build && cd build && cmake -GNinja -DCMAKE_TOOLCHAIN_FILE=${HOME}/toolchain -DFIPS=1 -DCMAKE_BUILD_TYPE=Release ..
 ninja
-ninja run_tests
+# Fails SSL test due to functional change.
+# ninja run_tests
```
Comment (RyanTheOptimist) [inline bazel/external/boringssl_fips.genrule_cmd:125]: Yeah, I think we'd definitely prefer to hold off on this if can.

**Review by RyanTheOptimist — COMMENTED** [review]
Looks good! Thanks for doing this. One nit and a recommendation to hold off landing this, if we can.

File: bazel/boringssl_fips.patch (hunk 1 above)
Comment (kyessenov) [reply bazel/boringssl_fips.patch:48]: I reverse engineered `set_compliance` method from the later boringssl - this method exists in new boringssl and is somewhat similar.

File: bazel/external/boringssl_fips.genrule_cmd (hunk 2 above)
Comment (kyessenov) [reply bazel/external/boringssl_fips.genrule_cmd:125]: Yeap, the failing tests all have to do with CHACHA20 - not sure what we can do about it really. Even if I narrow it down to a select few, still a risk.

**Comment by RyanTheOptimist** [comment]
/wait since we're not sure if we're going to land this

**Comment by kyessenov** [comment]
Our internal decision was to not to use this due to the unknown risks of modifying TLS code in boringssl.

**Pull request closed without merging**